apps/prairielearn/src/ee/models/enrollment.sql

```diff
@@ -0,0 +1,55 @@
+-- BLOCK insert_enrollment
+INSERT INTO
+  enrollments (user_id, course_instance_id)
+VALUES
+  ($user_id, $course_instance_id)
+RETURNING
+  *;
+
+-- BLOCK select_enrollment_for_user_in_course_instance
+SELECT
+  *
+FROM
+  enrollments
+WHERE
+  user_id = $user_id
+  AND course_instance_id = $course_instance_id;
+
+-- BLOCK select_enrollment_counts
+SELECT
+  (
+    COUNT(*) FILTER (
+      WHERE
+        pg.id IS NULL
+    )
+  )::integer AS free,
+  (
+    COUNT(*) FILTER (
+      WHERE
+        pg.id IS NOT NULL
+    )
+  )::integer AS paid
+FROM
+  enrollments AS e
+  JOIN course_instances AS ci ON (e.course_instance_id = ci.id)
+  JOIN pl_courses AS c ON (ci.course_id = c.id)
+  JOIN institutions AS i ON (i.id = c.institution_id)
+  LEFT JOIN plan_grants AS pg ON (
+    pg.institution_id = i.id
+    AND pg.course_instance_id = ci.id
+    AND pg.enrollment_id = e.id
+    AND pg.plan_name = 'basic'
+  )
+WHERE
+  (
+    $created_since::interval IS NULL
+    OR e.created_at > now() - $created_since::interval
+  )
+  AND (
+    $institution_id::bigint IS NULL
+    OR i.id = $institution_id
+  )
+  AND (
+    $course_instance_id::bigint IS NULL
+    OR ci.id = $course_instance_id
+  );
```

Comment:
This was originally two queries that were identical except for the `WHERE` conditions. I combined them into one; the behavior changes depending on which params are set. Let me know if you'd prefer to see two separate queries.

Reply:
I'm worried that this is going to use a seq scan because of those OR queries. As @jonatanschroeder discovered, postgres finds it very hard to use indexes with OR conditions. I'd rather split this into separate queries each of which has hard-coded WHERE conditions.

I also think we should add an index on `(institution_id, created_at)`.

Reply:
Fair, I can split it up. This is low-level code that shouldn't change much, so I'm fine with the duplication.

`enrollments.institution_id` isn't a column - did you mean `(course_instance_id, created_at)`? That would likely help: when running a version of this query that doesn't include the `OR` clauses, I see we're doing a `Parallel Seq Scan on enrollments e` to filter rows based on `created_at`. I can't tell for sure if the query planner would use that index, however:


```
 Aggregate  (cost=5818.50..5818.51 rows=1 width=8) (actual time=37.712..40.606 rows=1 loops=1)
   ->  Hash Left Join  (cost=1320.18..5818.11 rows=77 width=8) (actual time=1.617..40.433 rows=1825 loops=1)
         Hash Cond: ((i.id = pg.institution_id) AND (ci.id = pg.course_instance_id) AND (e.user_id = pg.user_id))
         ->  Nested Loop  (cost=1311.96..5809.29 rows=77 width=24) (actual time=1.605..40.008 rows=1825 loops=1)
               ->  Seq Scan on institutions i  (cost=0.00..1.30 rows=1 width=8) (actual time=0.012..0.029 rows=1 loops=1)
                     Filter: (id = 3)
                     Rows Removed by Filter: 23
               ->  Gather  (cost=1311.96..5807.22 rows=77 width=24) (actual time=1.590..39.743 rows=1825 loops=1)
                     Workers Planned: 1
                     Workers Launched: 1
                     ->  Nested Loop  (cost=311.96..4799.52 rows=45 width=24) (actual time=1.541..32.953 rows=912 loops=2)
                           ->  Hash Join  (cost=311.69..4752.16 rows=105 width=24) (actual time=1.519..29.765 rows=1064 loops=2)
                                 Hash Cond: (e.course_instance_id = ci.id)
                                 ->  Parallel Seq Scan on enrollments e  (cost=0.00..4440.20 rows=105 width=16) (actual time=0.347..28.260 rows=1064 loops=2)
                                       Filter: (created_at > (now() - '1 year'::interval))
                                       Rows Removed by Filter: 121564
                                 ->  Hash  (cost=279.64..279.64 rows=2564 width=16) (actual time=1.112..1.113 rows=2564 loops=2)
                                       Buckets: 4096  Batches: 1  Memory Usage: 153kB
                                       ->  Seq Scan on course_instances ci  (cost=0.00..279.64 rows=2564 width=16) (actual time=0.014..0.711 rows=2564 loops=2)
                           ->  Index Scan using pl_courses_pkey on pl_courses c  (cost=0.28..0.45 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=2129)
                                 Index Cond: (id = ci.course_id)
                                 Filter: (institution_id = 3)
                                 Rows Removed by Filter: 0
         ->  Hash  (cost=8.20..8.20 rows=1 width=32) (actual time=0.007..0.008 rows=0 loops=1)
               Buckets: 1024  Batches: 1  Memory Usage: 8kB
               ->  Index Scan using plan_grants_institution_course_instance_user_name_idx on plan_grants pg  (cost=0.15..8.20 rows=1 width=32) (actual time=0.006..0.006 rows=0 loops=1)
                     Index Cond: ((institution_id = 3) AND (plan_name = 'basic'::text))
 Planning Time: 1.008 ms
 Execution Time: 40.703 m
```

As written (but without the ORs), I might have expected the query plan to first find all enrollments for all course instances, and only then filter 🤔

Reply:
Talked about this with Matt - I shouldn't have doubted the query planner! I was running this query with UIUC enrollments, so PG correctly identified that most of the course instances and/or enrollments are in this institution, so it figured a sequential scan would be fastest. When running on an institution with many fewer course instances and enrollments, it avoided a sequential scan on `enrollments` and ran very fast (<1ms).

We decided that 40ms is acceptable for the time being, as this is comparable to queries that we're running on every request, not just generally once per user/course at the start of a semester.

Reply:
Are we actually using the query by `institution_id` _without_ a time restriction? If not, I'd suggest removing the OR in that query and requiring a time interval. If we do need the version with `created_since == null` then it's fine to leave it as is.

Reply:
Sure, I can make that a mandatory part of the query.